nvidia_gpu_stats/src/gpu_nvidia.rs

```diff
@@ -0,0 +1,252 @@
+            nvml,
+            cuda_version: format!(
+                "{}.{}",
+                nvml_wrapper::cuda_driver_version_major(cuda_version),
+                nvml_wrapper::cuda_driver_version_minor(cuda_version)
+            ),
+            device_count,
+        })
+    }
+
+    /// Function to check if a GPU is being used by a specific process or its children
+    fn gpu_in_use_by_process(&self, device: &Device, pid: i32) -> bool {
+        let our_pids: Vec<i32> = std::iter::once(pid)
+            .chain(self.get_child_pids(pid))
+            .collect();
+
+        let compute_processes = device.running_compute_processes().unwrap_or_default();
+        let graphics_processes = device.running_graphics_processes().unwrap_or_default();
+
+        let device_pids: Vec<i32> = compute_processes
+            .iter()
+            .chain(graphics_processes.iter())
+            .map(|p| p.pid as i32)
+            .collect();
+
+        our_pids.iter().any(|&p| device_pids.contains(&p))
+    }
+
+    /// Function to get child process IDs for a given parent PID
+    fn get_child_pids(&self, pid: i32) -> Vec<i32> {
+        let mut sys = System::new_all();
+        sys.refresh_all();
+
+        sys.processes()
+            .values()
+            .filter(|process| process.parent() == Some(Pid::from(pid as usize)))
+            .map(|process| process.pid().as_u32() as i32)
+            .collect()
+    }
+
+    /// Samples GPU metrics using NVML.
+    ///
+    /// This function collects various metrics from all available GPUs, including
+    /// utilization, memory usage, temperature, and power consumption. It also
+    /// checks if the specified process is using each GPU and collects process-specific
+    /// metrics if applicable.
+    ///
+    /// # Arguments
+    ///
+    /// * `metrics` - A mutable reference to a `Metrics` struct to store the collected metrics.
+    /// * `pid` - The process ID to monitor for GPU usage.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` with an empty tuple on success or an `NvmlError` on failure.
+    /// The collected metrics are stored in the `Metrics` struct provided as an argument.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    /// * NVML fails to retrieve device information
+    /// * Any of the metric collection operations fail
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crate::gpu_nvidia::NvidiaGpu;
+    /// use crate::metrics::Metrics;
+    /// let nvidia_gpu = NvidiaGpu::new().unwrap();
+    /// let mut metrics = Metrics::new();
+    /// nvidia_gpu.sample_metrics(&mut metrics, 1234).unwrap();
+    /// ```
+    pub fn sample_metrics(&self, metrics: &mut Metrics, pid: i32) -> Result<(), NvmlError> {
+        metrics.add_metric("cuda_version", &*self.cuda_version);
+        metrics.add_metric("_gpu.count", self.device_count);
+
+        for di in 0..self.device_count {
+            let device = self.nvml.device_by_index(di)?;
```

Comment:
If there's a failure for one device, we will return an error rather than skipping it. That's probably not what we want, right?

Reply:
good point, fixed